data/pools/transactionPool.go

```diff
@@ -892,14 +973,54 @@ func (pool *TransactionPool) assembleEmptyBlock(round basics.Round) (assembled *
 		return nil, err
 	}
 	next := bookkeeping.MakeBlock(prev)
-	blockEval, err := pool.ledger.StartEvaluator(next.BlockHeader, 0)
+	blockEval, err := pool.ledger.StartEvaluator(next.BlockHeader, 0, pool.calculateMaxTxnBytesPerBlock(next.BlockHeader.CurrentProtocol))
 	if err != nil {
 		err = fmt.Errorf("TransactionPool.assembleEmptyBlock: cannot start evaluator for %d: %v", round, err)
 		return nil, err
 	}
 	return blockEval.GenerateBlock()
 }
 
+// SetDataExchangeRate updates the data exchange rate this node is expected to have.
+func (pool *TransactionPool) SetDataExchangeRate(dataExchangeRate uint64) {
+	atomic.StoreUint64(&pool.latestMeasuredDataExchangeRate, dataExchangeRate)
+}
+
+// calculateMaxTxnBytesPerBlock computes the optimal block size for the current node, based
+// on it's effective network capabilities. This number is bound by the protocol MaxTxnBytesPerBlock.
+func (pool *TransactionPool) calculateMaxTxnBytesPerBlock(consensusVersion protocol.ConsensusVersion) int {
```

Comment:
Why can't we have a setter for the value that this function returns instead of `SetDataExchangeRate()`? It seems strange that transaction pool has anything to do with data rate.

Reply:
Ideally, it would be the network library that would have that value. The block evaluator would use the inputs from the network package to determine the bandwidth, and pass it right along to the block evaluator.

I think it's a good observation, but not something we need to have for this release. Subsequent PRs could easily accomplish that.

Reply:
Why should the block evaluator have bandwidth? I think we should just do this calculation where we call `SetDataExchangeRate()`.

Reply:
That sounds like a good idea to me. Let's create a ticket for that ?

Reply:
Why not change it now?

Reply:
I'm trying to minimize the amount of changes.. until we'll have enough feedback. The change could introduce more bugs - so I'm trying not to change anything ( unless we can find a real issue ).

Reply:
Ok, should I make a ticket?

Reply:
yes, please.

Reply:
https://github.com/algorand/go-algorand/issues/2754